### PR TITLE
Ctskf 870 special preparation fee

### PR DESCRIPTION
**Pull request by AlexRyanShep**
**This ticket is not to be merged to master, it will be merged into a feature branch manually once reviewed**

#### What
This ticket modifies the LGFS Misc Fees page to make an API call to Fee Calculator to retrieve Special Preparation fee rates

There are still a few unresolved issues around validation and some other feature tests broken by the changes, but due to project scope/time limit I'm having a call it here, as this is fulfilling the scope of being a PoC for making api calls from this page and getting valid responses. 

#### Ticket

[CTSKF-870](https://dsdmoj.atlassian.net/browse/CTSKF-870)

[CTSKF-870]: https://dsdmoj.atlassian.net/browse/CTSKF-870?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: spec/requests/prices_spec.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
+require 'rails_helper'
+
+RSpec.describe 'calcuate prices' do
+  describe 'POST /external_users/claims/claim_id/fees/calculate_price.json' do
+    subject(:calculate_price) { post external_users_claim_fees_calculate_price_url(claim.id, format: :json), params: }
+
+
```
Comment (mpw5) [inline spec/requests/prices_spec.rb]: There's probably no need for the extra whitespace here.

File: app/services/claims/fee_calculator/unit_price.rb [hunk 2]
```diff
@@ -44,8 +44,14 @@ def price_options
...
         opts[:unit] = unit
         opts.keep_if { |_k, v| v.present? }
+        # Filter above will filter out false, which is a valid value for london_rates_apply
+        opts[:london_rates_apply] = @london_rates_apply unless @london_rates_apply.nil?
+        opts
+
       end
 
+
+
```
Comment (mpw5) [inline app/services/claims/fee_calculator/unit_price.rb]: whitespace

File: app/services/claims/fee_calculator/unit_price.rb [hunk 3]
```diff
@@ -44,8 +44,14 @@ def price_options
         opts[:limit_to] = limit_to
         opts[:unit] = unit
         opts.keep_if { |_k, v| v.present? }
+        # Filter above will filter out false, which is a valid value for london_rates_apply
+        opts[:london_rates_apply] = @london_rates_apply unless @london_rates_apply.nil?
+        opts
+
```
Comment (mpw5) [inline app/services/claims/fee_calculator/unit_price.rb]: whitespace

File: features/fee_calculator/litigator/misc_fee_calculator_london.feature [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      | section       | fee_description                                    | rate   | hint                            | help |
+      | miscellaneous | Special Preparation fee                            | 43.12  | Number of hours                 | true |
+
+    Then I eject the VCR cassette
+    And I save as draft
+
+    Then I should see 'Draft claim saved'
+    And Claim 'A20161298' should be listed with a status of 'Draft' and a claimed amount of '£574.60'
+
+  Scenario: I create a fee scheme 10 fixed fee claim using calculated value
```
Comment (mpw5) [inline features/fee_calculator/litigator/misc_fee_calculator_london.feature]: Should this be `Scenario: I create a fee scheme 10 fixed fee claim using london rates` to match the scheme 9 scenario on line 5?

File: features/fee_calculator/litigator/misc_fee_calculator_nonlondon.feature [hunk 5]
```diff
@@ -0,0 +1,95 @@
+@javascript
+@fee_calc_vcr
+Feature: litigator completes fixed fee page using calculator
+
+  Scenario: I create a fee scheme 9 fixed fee claim using london rates
```
Comment (mpw5) [inline features/fee_calculator/litigator/misc_fee_calculator_nonlondon.feature]: Should this be `non-london` rates?

File: features/fee_calculator/litigator/misc_fee_calculator_nonlondon.feature [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      | section       | fee_description                                    | rate   | hint                            | help |
+      | miscellaneous | Special Preparation fee                            | 41.06  | Number of hours                 | true |
+
+    Then I eject the VCR cassette
+    And I save as draft
+
+    Then I should see 'Draft claim saved'
+    And Claim 'A20161298' should be listed with a status of 'Draft' and a claimed amount of '£567.18'
+
+  Scenario: I create a fee scheme 10 fixed fee claim using calculated value
```
Comment (mpw5) [inline features/fee_calculator/litigator/misc_fee_calculator_nonlondon.feature]: As above, consider changing this to match the wording of the scheme 9 scenario.

File: app/services/claims/fee_calculator/calculate_price.rb [hunk 7]
```diff
@@ -137,7 +137,10 @@ def fee_type_mappings
       end
 
       def fee_type_code_for(fee_type)
-        return 'LIT_FEE' if lgfs?
+
```
Comment (mpw5) [inline app/services/claims/fee_calculator/calculate_price.rb]: whitespace

File: app/views/external_users/claims/misc_fees/litigators/_misc_fee_fields.html.haml [hunk 8]
```diff
@@ -1,22 +1,55 @@
...
+      claim.eligible_misc_fee_type_options_for_select.map{ |ft| [ft.first, ft.second, { 'data-unique-code': ft.third[:data][:unique_code] }] },
+      :second,
+      :first,
+      class: 'js-misc-fee-type js-fee-type js-fee-calculator-fee-type fx-misc-fee-calculation',
+      form_group: { class: 'fee-type fx-autocomplete-wrapper cc-fee-type' },
+      label: { text: t('.fee_type_html', context: t('.misc_fee')), size: 's' },
+      options: { include_blank: '' }
+
+    -# - if misc_fee_type.unique_code.eql?('MIUMO')
+    -#   = render 'warnings/unused_material_over_3_hours'
```
Comment (mpw5) [inline app/views/external_users/claims/misc_fees/litigators/_misc_fee_fields.html.haml]: Is this needed? Should it be commented out?

File: features/page_objects/sections/lgfs_misc_fee_section.rb [hunk 9]
```diff
@@ -3,6 +3,7 @@
 class FeeTypeSection < SitePrism::Section
   sections :radios, RadioSection, '.govuk-radios__item'
 
+
```
Comment (mpw5) [inline features/page_objects/sections/lgfs_misc_fee_section.rb]: whitespace

**Review by mpw5 — APPROVED** [review]
LGTM. Minor/cosmetic feedback but nothing to worry about now 🏅

File: spec/requests/prices_spec.rb (hunk 1 above)
Comment (AlexRyanShep) [reply spec/requests/prices_spec.rb]: This was picked up in a later commit 👍

File: app/services/claims/fee_calculator/unit_price.rb (hunk 2 above)
Comment (AlexRyanShep) [reply app/services/claims/fee_calculator/unit_price.rb]: As above 🙂

File: app/services/claims/fee_calculator/unit_price.rb (hunk 3 above)
Comment (AlexRyanShep) [reply app/services/claims/fee_calculator/unit_price.rb]: As above 🙂

File: features/fee_calculator/litigator/misc_fee_calculator_london.feature (hunk 4 above)
Comment (AlexRyanShep) [reply features/fee_calculator/litigator/misc_fee_calculator_london.feature]: Good catch thanks! Not sure how that happened

File: features/fee_calculator/litigator/misc_fee_calculator_nonlondon.feature (hunk 5 above)
Comment (AlexRyanShep) [reply features/fee_calculator/litigator/misc_fee_calculator_nonlondon.feature]: 👍

File: features/fee_calculator/litigator/misc_fee_calculator_nonlondon.feature (hunk 6 above)
Comment (AlexRyanShep) [reply features/fee_calculator/litigator/misc_fee_calculator_nonlondon.feature]: 👍  Thanks! I think these were a mixture of tiredness and me using one of the existing litigator tests as a base

File: app/services/claims/fee_calculator/calculate_price.rb (hunk 7 above)
Comment (AlexRyanShep) [reply app/services/claims/fee_calculator/calculate_price.rb]: 👍

File: app/views/external_users/claims/misc_fees/litigators/_misc_fee_fields.html.haml (hunk 8 above)
Comment (AlexRyanShep) [reply app/views/external_users/claims/misc_fees/litigators/_misc_fee_fields.html.haml]: This was fixed in a commit after the one you've commented on 🙂

File: features/page_objects/sections/lgfs_misc_fee_section.rb (hunk 9 above)
Comment (AlexRyanShep) [reply features/page_objects/sections/lgfs_misc_fee_section.rb]: 👍

**Comment by AlexRyanShep** [comment]
@mpw5 Made a new commit with some changes based on your comments 🙂

**Comment by AlexRyanShep** [comment]
Closing and merging to feature branch

**Pull request closed without merging**